In [85]:
import pandas as pd

# Работа с Pandas

1. Загрузите ваш DataFrame в переменную df и выведете на экран первые пять строк 

In [86]:
df = pd.read_csv('../data/googleplaystore .csv')

In [87]:
df.head()

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0.0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0.0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,"5,000,000+",Free,0.0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0.0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,"100,000+",Free,0.0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up


2. Отобразите количество строк и столбцов в DataFrame

In [88]:
df.shape

(10841, 13)

3. Выведете основную информацию о DataFrame методом info(). Проанализируйте есть ли пропуски в столбцах. Напишите вывод о том, какие столбцы содержат пропуски

In [89]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  str    
 1   Category        10841 non-null  str    
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  str    
 4   Size            10841 non-null  str    
 5   Installs        10841 non-null  str    
 6   Type            10840 non-null  str    
 7   Price           10841 non-null  float64
 8   Content Rating  10840 non-null  str    
 9   Genres          10841 non-null  str    
 10  Last Updated    10841 non-null  str    
 11  Current Ver     10833 non-null  str    
 12  Android Ver     10838 non-null  str    
dtypes: float64(2), str(11)
memory usage: 1.1 MB


## Нули в:
- `Rating`: float64
- `Type`: str
- `Content Rating`: str
- `Current Ver`: str
- `Android Ver`: str

4. Удалите или замените пропуски известными вам методами. (Подсказка: `dropna()` или `fillna()`, выбор того или иного метода должен быть аргументирован).

In [90]:
print(df['Rating'].mean(), df['Rating'].median())

4.193338315362443 4.3


In [91]:
df['Rating'] = df['Rating'].fillna(df['Rating'].median())

In [92]:
df.describe(include='string')

,App,Category,Reviews,Size,Installs,Type,Content Rating,Genres,Last Updated,Current Ver,Android Ver
count,10841,10841,10841,10841,10841,10840,10840,10841,10841,10833,10838
unique,9660,34,6002,462,22,3,6,120,1378,2832,33
top,ROBLOX,FAMILY,0,Varies with device,"1,000,000+",Free,Everyone,Tools,"August 3, 2018",Varies with device,4.1 and up
freq,9,1972,596,1695,1579,10039,8714,842,326,1459,2451


In [93]:
df['Type'] = df['Type'].fillna('Free')

In [94]:
df['Content Rating'] = df['Content Rating'].fillna('Everyone')

In [95]:
df['Current Ver'] = df['Current Ver'].fillna('Varies with device')

In [96]:
df['Android Ver'] = df['Android Ver'].fillna('4.1 and up')

5. Проверьте нет ли пропусков после произведенной обработки

In [97]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  str    
 1   Category        10841 non-null  str    
 2   Rating          10841 non-null  float64
 3   Reviews         10841 non-null  str    
 4   Size            10841 non-null  str    
 5   Installs        10841 non-null  str    
 6   Type            10841 non-null  str    
 7   Price           10841 non-null  float64
 8   Content Rating  10841 non-null  str    
 9   Genres          10841 non-null  str    
 10  Last Updated    10841 non-null  str    
 11  Current Ver     10841 non-null  str    
 12  Android Ver     10841 non-null  str    
dtypes: float64(2), str(11)
memory usage: 1.1 MB


6. Выведите на экран минимум и максимум из столбца `Price`:

In [98]:
print(f'Минимум столбца `Price` - {df['Price'].min()},\nМаксимум столбца `Price` - {df['Price'].max()}')

Минимум столбца `Price` - 0.0,
Максимум столбца `Price` - 400.0


7. Выведите на экран медиану и среднее арифметическое столбцов:
 - Rating
 - Reviews

In [99]:
df['Reviews'].sample(20)

603            59
6986        51791
1991      1480182
9975       432160
3066        32386
9368           18
5664        14657
2680        30834
557         17268
4887          533
4325       357049
2454            4
9563         4254
4685       430643
10458        2390
5075       152653
1842     22429716
556        222888
559          4953
2393         4303
Name: Reviews, dtype: str

In [100]:
df[df['Reviews'].astype(str).str.contains(r'M', regex=True)]

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
10472,Life Made WI-Fi Touchscreen Photo Frame,1.9,19.0,3.0M,"1,000+",Free,0,0.0,Everyone,"February 11, 2018",1.0.19,4.0 and up,4.1 and up


In [101]:
df['Reviews'] = df['Reviews'].replace('3.0M', '3000000')

In [102]:
df['Reviews'] = df['Reviews'].astype(float)

In [103]:
cols = ['Rating', 'Reviews']

print('=======================')
print()
for col in cols:
    print(f'Столбец {col}')
    print(f'Среднее значение в {col} - {df[col].mean()}')
    print(f'Медианное значение в {col} - {df[col].median()}')
    
    print()
    print('=======================')
    print()


Столбец Rating
Среднее значение в Rating - 4.2078406051102295
Медианное значение в Rating - 4.3


Столбец Reviews
Среднее значение в Reviews - 444388.6535374965
Медианное значение в Reviews - 2094.0




8. Выведите на экран все уникальные значения категориального столбца Genres

In [104]:
for name in df['Genres'].unique():
    print(name)

Art & Design
Art & Design;Pretend Play
Art & Design;Creativity
Art & Design;Action & Adventure
Auto & Vehicles
Beauty
Books & Reference
Business
Comics
Comics;Creativity
Communication
Dating
Education;Education
Education
Education;Creativity
Education;Music & Video
Education;Action & Adventure
Education;Pretend Play
Education;Brain Games
Entertainment
Entertainment;Music & Video
Entertainment;Brain Games
Entertainment;Creativity
Events
Finance
Food & Drink
Health & Fitness
House & Home
Libraries & Demo
Lifestyle
Lifestyle;Pretend Play
Adventure;Action & Adventure
Arcade
Casual
Card
Casual;Pretend Play
Action
Strategy
Puzzle
Sports
Music
Word
Racing
Casual;Creativity
Casual;Action & Adventure
Simulation
Adventure
Board
Trivia
Role Playing
Simulation;Education
Action;Action & Adventure
Casual;Brain Games
Simulation;Action & Adventure
Educational;Creativity
Puzzle;Brain Games
Educational;Education
Card;Brain Games
Educational;Brain Games
Educational;Pretend Play
Entertainment;Education
Ca

9. Сгруппируйте данные по столбцу `Genres` и посчитайте для каждого жанра средний и медианный рейтинг. Представьте результат в виде нового DataFrame,который будет сохранен в переменную `grouped_df`, где одна колонка будет содержать названия жарнов, а две другие буду содержать в себе средний и медианный рейтинги. Напишите краткий вывод.

In [105]:
grouped_df = df.groupby('Genres', as_index=False).agg(
    mean_rating = ('Rating', 'mean'),
    median_rating = ('Rating', 'median'),
)

In [106]:
grouped_df.head()

,Genres,mean_rating,median_rating
0,Action,4.285753,4.3
1,Action;Action & Adventure,4.311765,4.3
2,Adventure,4.184000,4.3
3,Adventure;Action & Adventure,4.423077,4.5
4,Adventure;Brain Games,4.600000,4.6


In [107]:
grouped_df.describe()

,mean_rating,median_rating
count,120.000000,120.000000
mean,4.390470,4.436250
std,1.355493,1.350598
min,3.800000,3.800000
25%,4.179975,4.300000
50%,4.281539,4.300000
75%,4.351724,4.400000
max,19.000000,19.000000


- Средние рейтинги и медианные по жанрам приблизительно совпадают;
- Минимальные и максимальные значения совпадают полностью.

10. Из получившегося датафрейма выведите на экран только те жанры, чей медианный рейтинг больше 4.5. Напишите краткий вывод.

In [108]:
grouped_df[grouped_df['median_rating'] > 4.5]

,Genres,mean_rating,median_rating
4,Adventure;Brain Games,4.600000,4.6
11,Art & Design;Creativity,4.400000,4.7
18,Board;Pretend Play,4.800000,4.8
35,Comics;Creativity,4.800000,4.8
55,Entertainment;Creativity,4.533333,4.6
60,"February 11, 2018",19.000000,19.0
65,Health & Fitness;Education,4.700000,4.7
75,Music;Music & Video,4.533333,4.6
88,Puzzle;Education,4.600000,4.6
106,Strategy;Action & Adventure,4.600000,4.6


- Рейтинги по прежнему приблизительно одинаковые;
- Из 120 наблюдений рейтингом более 4.5 обладают всего 9 фильмов;
- Наблюдение с индексом `60` скорее всего является ошибочным и подлежит удалению, тк на месте жанра там дата, а средний и медианный значения рейтинга являются аномальными значениями для поля `Rating`.  

In [109]:
df[df['Rating'] > 10]

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
10472,Life Made WI-Fi Touchscreen Photo Frame,1.9,19.0,3000000.0,"1,000+",Free,0,0.0,Everyone,"February 11, 2018",1.0.19,4.0 and up,4.1 and up


In [110]:
df = df.drop(index=10472)

In [111]:
df.describe()

,Rating,Reviews,Price
count,10840.000000,1.084000e+04,10840.000000
mean,4.206476,4.441529e+05,1.027368
std,0.480342,2.927761e+06,15.949703
min,1.000000,0.000000e+00,0.000000
25%,4.100000,3.800000e+01,0.000000
50%,4.300000,2.094000e+03,0.000000
75%,4.500000,5.477550e+04,0.000000
max,5.000000,7.815831e+07,400.000000


11. Выведете кол-во дубликатов в основном DataFrame (df) и удалите их 

In [112]:
df.duplicated().sum()

np.int64(483)

In [113]:
df = df.drop_duplicates()

12. Посчитатайте какое количество приложений содержится в каждом жанре (value_counts).В получившемся Series выведите на экран 10 строк с наибольшим количеством приложений отсортированных по убыванию.

In [116]:
df['Genres'].value_counts().head(10)

Genres
Tools              842
Entertainment      588
Education          527
Business           427
Medical            408
Productivity       407
Personalization    388
Lifestyle          372
Communication      366
Sports             364
Name: count, dtype: int64